File: server/src/services/bankingClient.js
import axios from 'axios';

const BANKS = {
  vbank: 'https://vbank.open.bankingapi.ru',
  abank: 'https://abank.open.bankingapi.ru',
  sbank: 'https://sbank.open.bankingapi.ru'
};

class BankingClient {
  constructor() {
    this.teamId = process.env.OPEN_BANKINGAPI_TEAM_ID || 'team096';
    this.teamSecret = process.env.OPEN_BANKINGAPI_PASSWORD || '';
    this.tokens = new Map(); // bank -> { token, expiresAt }
  }

  /**
   * Получить токен для банка
   */
  async getBankToken(bank) {
    const bankUrl = BANKS[bank];
    if (!bankUrl) {
      throw new Error(`Unknown bank: ${bank}`);
    }

    // Проверяем кэш токена
    const cached = this.tokens.get(bank);
    if (cached && cached.expiresAt > Date.now()) {
      return cached.token;
    }

    try {
      const response = await axios.post(
        `${bankUrl}/auth/bank-token`,
        null,
        {
          params: {
            client_id: this.teamId,
            client_secret: this.teamSecret
          }
        }
      );

      const { access_token, expires_in } = response.data;
      const expiresAt = Date.now() + (expires_in - 60) * 1000; // -60 секунд для запаса

      this.tokens.set(bank, { token: access_token, expiresAt });
      return access_token;
    } catch (error) {
      console.error(`Error getting token for ${bank}:`, error.response?.data || error.message);
      throw new Error(`Failed to get token for ${bank}: ${error.response?.data?.detail || error.message}`);
    }
  }

  /**
   * Сделать запрос к банку
   */
  async request(bank, method, endpoint, options = {}) {
    const bankUrl = BANKS[bank];
    if (!bankUrl) {
      throw new Error(`Unknown bank: ${bank}`);
    }

    const token = await this.getBankToken(bank);
    const url = `${bankUrl}${endpoint}`;

    const config = {
      method,
      url,
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
        ...options.headers
      }
    };
    
    if (options.params) {
      config.params = options.params;
    }
    if (options.data) {
      config.data = options.data;
    }

    try {
      const response = await axios(config);
      return response.data;
    } catch (error) {
      console.error(`Error ${method} ${url}:`, error.response?.data || error.message);
      throw {
        status: error.response?.status || 500,
        message: error.response?.data?.detail || error.message,
        data: error.response?.data
      };
    }
  }

  /**
   * Получить список всех банков
   */
  getBanks() {
    return Object.keys(BANKS);
  }

  /**
   * Получить URL банка
   */
  getBankUrl(bank) {
    return BANKS[bank];
  }
}

export default new BankingClient();

File: server_fastapi/services/banking_client.py
import httpx
import logging
from typing import Dict, Optional
from datetime import datetime, timedelta

from config import settings

logger = logging.getLogger(__name__)

BANKS = {
    "vbank": "https://vbank.open.bankingapi.ru",
    "abank": "https://abank.open.bankingapi.ru",
    "sbank": "https://sbank.open.bankingapi.ru"
}


class BankingClient:
    def __init__(self):
        self.team_id = settings.open_bankingapi_team_id
        self.team_secret = settings.open_bankingapi_password
        self.tokens: Dict[str, Dict] = {}  # bank -> { token, expires_at }
    
    async def get_bank_token(self, bank: str) -> str:
        """Get access token for bank"""
        bank_url = BANKS.get(bank)
        if not bank_url:
            raise ValueError(f"Unknown bank: {bank}")
        
        # Check token cache
        cached = self.tokens.get(bank)
        if cached and cached["expires_at"] > datetime.now():
            return cached["token"]
        
        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{bank_url}/auth/bank-token",
                    params={
                        "client_id": self.team_id,
                        "client_secret": self.team_secret
                    }
                )
                response.raise_for_status()
                data = response.json()
                
                access_token = data.get("access_token")
                expires_in = data.get("expires_in", 3600)
                
                # Subtract 60 seconds for safety margin
                expires_at = datetime.now() + timedelta(seconds=expires_in - 60)
                
                self.tokens[bank] = {
                    "token": access_token,
                    "expires_at": expires_at
                }
                
                return access_token
        except httpx.HTTPError as e:
            logger.error(f"Error getting token for {bank}: {e}")
            raise Exception(f"Failed to get token for {bank}: {str(e)}")
    
    async def request(
        self,
        bank: str,
        method: str,
        endpoint: str,
        params: Optional[Dict] = None,
        data: Optional[Dict] = None,
        headers: Optional[Dict] = None
    ) -> Dict:
        """Make request to bank API"""
        bank_url = BANKS.get(bank)
        if not bank_url:
            raise ValueError(f"Unknown bank: {bank}")
        
        token = await self.get_bank_token(bank)
        url = f"{bank_url}{endpoint}"
        
        request_headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
            **(headers or {})
        }
        
        try:
            async with httpx.AsyncClient() as client:
                response = await client.request(
                    method=method.upper(),
                    url=url,
                    params=params,
                    json=data,
                    headers=request_headers,
                    timeout=30.0
                )
                response.raise_for_status()
                return response.json()
        except httpx.HTTPStatusError as e:
            logger.error(f"Error {method} {url}: {e.response.text}")
            raise {
                "status": e.response.status_code,
                "message": e.response.json().get("detail", str(e)),
                "data": e.response.json()
            }
        except httpx.HTTPError as e:
            logger.error(f"Error {method} {url}: {e}")
            raise {
                "status": 500,
                "message": str(e),
                "data": None
            }
    
    def get_banks(self) -> list:
        """Get list of all banks"""
        return list(BANKS.keys())
    
    def get_bank_url(self, bank: str) -> Optional[str]:
        """Get URL for bank"""
        return BANKS.get(bank)


# Singleton instance
banking_client = BankingClient()

